[20.6] Add optional potion fluid

ApexModder opened the pull request:
With the addition of data components and more specifically the PotionContents component type, I feel mods which added potion fluids would now all be adding duplicate fluids all doing the same thing, wrapping the PotionContents component into a fluid stack.

With this PR modders could simply request Neo to enable the potion fluid rather than needing to implement one themselves.
`NeoForgeMode.enablePotionFluid()`

This PR additionally (if the potion fluid is enabled)
- marks the vanilla Potion, Splash Potion and Lingering Potion as valid fluid containers.
- Registers a new bucket item for the potion fluid.

ApexModder committed:
Add optional potion fluid enabled via NeoForgeMod.enablePotionFluid

Technici4n reviewed (changes requested):
What about fluids for lingering and splash potions?

Also, please try not to add too much code to `NeoForgeMod`. The fluid handler and the fluid can be moved elsewhere at the very least.

KnightMiner commented:
In my opinion, splash potions and lingering potions should be the same fluid as regular potions, its just a different bottle. I would rather not see Neo force them being the same on me as it breaks some principles for how I interact with potions. I'm not the only one, pretty sure Create is the only mod that puts the potion bottle in the fluid (hence why they do not use the same potion tag as the rest of us).

----

In general, I really think Neo should have fewer fluids provided for you, not more. If it was up to me we would ditch Neo milk fluid and mods would just use a milk tag.

One notable problem this PR will cause with me is I lose the ability to make potion buckets drinkable, which was a neat future I could do by reusing the PotionItem class. I'd honestly rather you just register the data components for me to use with my own custom potion fluid (which looking through this PR seems unneeded as you just reuse the vanilla component, which is the same thing mods were doing before data components was a thing). Plus, forcing the texture to be just colored water is a loss as well.

ApexModder committed:
Extract code out of NeoForgeMod

To allow modders to easily extend potion fluid logic for their own potion fluids

ApexModder committed:
Make potion bucket drinkable

ApexModder commented:
> Also, please try not to add too much code to NeoForgeMod. The fluid handler and the fluid can be moved elsewhere at the very least.

I have extracted out all potion related logic from NeoForgeMod into their own classes under `net.neoforged.neoforge.fluids.potion`, this should allow others to extend them for their own potion fluid implementations / use as reference.

> One notable problem this PR will cause with me is I lose the ability to make potion buckets drinkable,

I have also made the potion buckets drinkable, they had no right click functionality before as the potion fluid cannot be placed.

Technici4n commented on src/main/java/net/neoforged/neoforge/fluids/FluidType.java:
```diff
@@ -67,6 +67,11 @@ public class FluidType {
+    /**
+     * The number of fluid units that a bottle represents.
+     */
+    public static final int BOTTLE_VOLUME = BUCKET_VOLUME / 4;
```
Questionable. ^^

ApexModder replied:
The standard amount for bottles is 250mb (4 bottles per bucket)
the math here makes it clear its 4 bottles to 1 bucket (and does indeed result in 250)

TelepathicGrunt replied:
This is probably the only part of the PR that is good imo. As if you use 3 bottles rn, others in the community yells at you to switch to 4 because mods with recipes of 3 and mods with recipes of 4 clashes in ways that can cause fluid duping bugs when abusing both recipes. Standardizing that 4 bottles = 1 bucket for all non-vanilla water fluids would finally put the recipe conflicts to rest. Or at least, finally get people to stop changing their mind between 3 bottles or 4 bottles every few years.

TelepathicGrunt commented:
If neo making more registered fluids for mods to use by default becomes a norm and honey fluid gets added, I will be incredibly upset as it'll then be fighting my honey fluid, causing confusion, and putting pressure on me to delete my extremely unique honey fluid I poured a lot of work into.

I feel bad for anyone that had a potion fluid they worked on to making it super unique only to be gutted by this generic potion fluid from neo.

KnightMiner commented on src/main/java/net/neoforged/neoforge/fluids/potion/PotionBucketItem.java:
```diff
@@ -0,0 +1,109 @@
+            CriteriaTriggers.CONSUME_ITEM.trigger(sPlayer, stack);
+
+        if (!level.isClientSide) {
+            stack.getOrDefault(DataComponents.POTION_CONTENTS, PotionContents.EMPTY).forEachEffect(effectInstance -> {
```
When I said "I also made it drinkable", there are a lot of details that go into it being drinkable beyond just being a clone of the vanilla item. Notably, a bucket of potion holds 4x as much as a bottle, so it does not make sense that it acts identically to a potion bottle. However, how much more effect you get is up for debate, any value is going to be somewhat controversial.

I personally choose to make the effect 2.5x as strong as vanilla potions as that number behaved nicely with things like health potions and you get the advantage of longer durations reducing when you need to drink again. But that is just what I did, someone else might have a different opinion; I am not entirely convinced it should be Neo deciding what the choice is.

KnightMiner commented on src/main/java/net/neoforged/neoforge/fluids/potion/PotionBucketItem.java:
```diff
@@ -0,0 +1,109 @@
+
+    @Override
+    public int getUseDuration(ItemStack stack) {
+        return 32;
```
Another detail Neo would force that I changed in my implementation, drinking more potion I made take longer. Modders lose the option to choose here. I don't think making this all an API is the right call.

KnightMiner commented on src/main/java/net/neoforged/neoforge/fluids/potion/PotionBucketItem.java:
```diff
@@ -0,0 +1,109 @@
+            stack.consume(1, player);
+        }
+
+        if (player == null || !player.hasInfiniteMaterials()) {
+            if (!stack.isEmpty())
+                return new ItemStack(Items.BUCKET);
+            if (player != null)
+                player.getInventory().add(new ItemStack(Items.BUCKET));
+        }
```
Pretty sure this has a vanilla helper for it, in DrinkUtils IIRC. Though that might have been a slightly different case

KnightMiner commented on src/main/java/net/neoforged/neoforge/fluids/potion/PotionFluidHandlerItem.java:
```diff
@@ -0,0 +1,125 @@
+import net.neoforged.neoforge.fluids.capability.IFluidHandler;
+import net.neoforged.neoforge.fluids.capability.IFluidHandlerItem;
+
+// Copy of FluidBucketWrapper to allow differing capacities and empty containers
```
Instead of copying, why not either extend it or extract the shared logic into a base class? Duplicate code tends to be a bad idea if it can be avoided.

There was just one method I had to override in my version of this IIRC.

KnightMiner commented on src/main/java/net/neoforged/neoforge/fluids/potion/PotionFluidType.java:
```diff
@@ -0,0 +1,43 @@
+    @Override
+    public ItemStack getBucket(FluidStack stack) {
+        // if theres at least 1 bucket, we use the new potion bucket item
+        // otherwise return a potion bottle
```
That is not what this code is doing. This code says if there is an exact multiple of buckets use bucket, otherwise use potion.

ApexModder commented:
Closing due to users preferring neo not provide these kinds of fluids and would rather modders implement them as needed in their mods.
